p2p/discovery/broadcast.go

```diff
@@ -0,0 +1,187 @@
+package discovery
+
+import (
+	"context"
+	"math/rand"
+	"net"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/libp2p/go-libp2p-host"
+	"github.com/libp2p/go-libp2p-peer"
+	pstore "github.com/libp2p/go-libp2p-peerstore"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr-net"
+)
+
+type broadcastService struct {
+	host          host.Host
+	listenSock    *net.UDPConn   // Socket listening for broadcasts
+	sendSocks     []*net.UDPConn // Sockets that send broadcasts
+	lk            sync.Mutex
+	notifees      []Notifee
+	broadcastPort int
+	active        bool
+}
+
+//
+// Bind a socket to send and recv broadcasts
+//
+func bindBroadcast(ip net.IP, port int) (*net.UDPConn, error) {
+	conn, err := net.ListenUDP("udp4", &net.UDPAddr{IP: ip, Port: port})
+	if err == nil {
+		file, err := conn.File()
+		if err == nil {
+			err = syscall.SetsockoptInt(int(file.Fd()), syscall.SOL_SOCKET, syscall.SO_BROADCAST, 1)
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
+
+func NewBroadcastService(ctx context.Context, peerhost host.Host, defaultPort int, interval time.Duration) (Service, error) {
+
+	// Broadcasts can only be received from 0.0.0.0 or 255.255.255.255
+	// but they can not be sent from 255.255.255.255.
+	// Therefor one listening socket is bound to 255.255.255.255 and a
+	// sending socket will be bound to each address that broadcasts
+	// should be sent from.
+	listenSock, err := bindBroadcast(net.IPv4bcast, defaultPort)
+	if err != nil {
+		return nil, err
+	}
+
+	bs := &broadcastService{
+		host:          peerhost,
+		listenSock:    listenSock,
+		broadcastPort: defaultPort,
+		active:        true,
+	}
+
+	addrs, err := getDialableListenAddrs(peerhost)
+	if err != nil {
+		bs.Close()
+		return nil, err
+	} else {
+		for _, a := range addrs {
+			if a.IP.IsGlobalUnicast() {
+				// binding to the same address and port as the TCP service
+				// embeds that address and port into broadcast messages
+				otherSock, err := bindBroadcast(a.IP, a.Port)
+				if err == nil {
+					bs.sendSocks = append(bs.sendSocks, otherSock)
+				}
+			}
+		}
+	}
```

Comment:
We may want to return an error if we fail to bind to *anything*. I wouldn't stop IPFS from starting in that case but, if we return an error and *log* it, it'll make debugging easier.

Reply:
When I looked at the mdns file I saw three different methods of logging errors, so I thought it cleaner not to log at all.

Reply:
We usually use `github.com/ipfs/go-log`. Any other methods of logging should be fixed. However, I'd just suggest returning an error and logging it from go-ipfs.